File: Mechanica Engine/src/font/Line.java
package font;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents a line of text during the loading of a text.
 * 
 * @author Karl
 *
 */
public class Line {

	private double maxLength;
	private double spaceSize;
	private double fontSize;

	private Word[] savedWords = new Word[32];
	private int currentWord = 0;
	private List<Word> words = new ArrayList<Word>();
	private double currentLineLength = 0;

	protected Line() {
		for (int i = 0; i < savedWords.length; i++) {
			savedWords[i] = new Word();
		}
	}

	protected int getNumberOfWords(){
		return currentWord+1;
	}

	protected void set(double spaceWidth, double fontSize, double maxLength){
		currentWord = 0;
		currentLineLength = 0;
		this.spaceSize = spaceWidth * fontSize;
		this.maxLength = maxLength;
		for (int i = 0; i < savedWords.length; i++) {
			savedWords[i].setFontSize(fontSize);
		}
		savedWords[0].set();

	}

	protected void addWord(){
		currentWord++;
		savedWords[currentWord].set();
	}


	/**
	 * Creates an empty line.
	 * 
	 * @param spaceWidth
	 *            - the screen-space width of a space character.
	 * @param fontSize
	 *            - the size of font being used.
	 * @param maxLength
	 *            - the screen-space maximum length of a line.
	 */
	protected Line(double spaceWidth, double fontSize, double maxLength) {
		this.spaceSize = spaceWidth * fontSize;
		this.maxLength = maxLength;
	}

	/**
	 * Attempt to add a word to the line. If the line can fit the word in
	 * without reaching the maximum line length then the word is added and the
	 * line length increased.
	 * 
	 * @param word
	 *            - the word to try to add.
	 * @return {@code true} if the word has successfully been added to the line.
	 */
	protected boolean attemptToAddWord(Word word) {
		double additionalLength = word.getWordWidth();
		additionalLength += !words.isEmpty() ? spaceSize : 0;
		if (currentLineLength + additionalLength <= maxLength) {
			words.add(word);
			currentLineLength += additionalLength;
			return true;
		} else {
			return false;
		}
	}

	protected boolean attemptToAddWord() {
		double additionalLength = savedWords[currentWord].getWordWidth();
		additionalLength +=  currentWord==0? 0 : spaceSize;
		if (currentLineLength + additionalLength <= maxLength) {
			addWord();
			currentLineLength += additionalLength;
			return true;
		} else {
			return false;
		}
	}

	protected void setWord(int word){
		savedWords[word].set();
	}

	protected void setCurrentWord(){
		savedWords[currentWord].set();
	}


	protected Word getWord(int word){
		return savedWords[word];
	}

	protected Word getCurrentWord(){
		return savedWords[currentWord];
	}

	/**
	 * @return The max length of the line.
	 */
	protected double getMaxLength() {
		return maxLength;
	}

	/**
	 * @return The current screen-space length of the line.
	 */
	protected double getLineLength() {
		return currentLineLength;
	}

	protected void arrayToList(){
		for (int i = 0; i <= currentWord; i++) {
			words.add(savedWords[i]);
		}
	}
	/**
	 * @return The list of words in the line.
	 */
	protected List<Word> getWords() {
		return words;
	}

}

File: lwjgl3-test/src/font/Line.kt
package font

import java.util.ArrayList

/**
 * Represents a line of text during the loading of a text.
 *
 * @author Karl
 */
class Line() {

    /**
     * @return The max length of the line.
     */
    var maxLength: Double = 0.0
        private set
    private var spaceSize: Double = 0.0
    private val fontSize: Double = 0.0

    private val savedWords = Array<Word>(32, { Word() })
    private var currentWord = 0
    private val words = ArrayList<Word>()
    /**
     * @return The current screen-space length of the line.
     */
    var lineLength = 0.0
        private set

    val numberOfWords: Int
        get() = currentWord + 1


    operator fun set(spaceWidth: Double, fontSize: Double, maxLength: Double) {
        currentWord = 0
        lineLength = 0.0
        this.spaceSize = spaceWidth * fontSize
        this.maxLength = maxLength
        for (i in savedWords.indices) {
            savedWords[i].setFontSize(fontSize)
        }
        savedWords[0].set()

    }

    fun addWord() {
        currentWord++
        savedWords[currentWord].set()
    }


    /**
     * Creates an empty line.
     *
     * @param spaceWidth
     * - the screen-space width of a space character.
     * @param fontSize
     * - the size of font being used.
     * @param maxLength
     * - the screen-space maximum length of a line.
     */
    constructor(spaceWidth: Double, fontSize: Double, maxLength: Double):this() {
        this.spaceSize = spaceWidth * fontSize
        this.maxLength = maxLength
    }

    /**
     * Attempt to add a word to the line. If the line can fit the word in
     * without reaching the maximum line length then the word is added and the
     * line length increased.
     *
     * @param word
     * - the word to try to add.
     * @return `true` if the word has successfully been added to the line.
     */
    fun attemptToAddWord(word: Word): Boolean {
        var additionalLength = word.wordWidth
        additionalLength += if (!words.isEmpty()) spaceSize else 0.0
        if (lineLength + additionalLength <= maxLength) {
            words.add(word)
            lineLength += additionalLength
            return true
        } else {
            return false
        }
    }

    fun attemptToAddWord(): Boolean {
        var additionalLength = savedWords[currentWord].wordWidth
        additionalLength += if (currentWord == 0) 0.0 else spaceSize
        if (lineLength + additionalLength <= maxLength) {
            addWord()
            lineLength += additionalLength
            return true
        } else {
            return false
        }
    }

    fun setWord(word: Int) {
        savedWords[word].set()
    }

    fun setCurrentWord() {
        savedWords[currentWord].set()
    }


    fun getWord(word: Int): Word {
        return savedWords[word]
    }

    fun getCurrentWord(): Word {
        return savedWords[currentWord]
    }

    fun arrayToList() {
        for (i in 0..currentWord) {
            words.add(savedWords[i])
        }
    }

    /**
     * @return The list of words in the line.
     */
    fun getWords(): List<Word> {
        return words
    }

}
